File: motor3d 2.0.ConvertedToC#/ElementoZBuffer.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Xml.Linq;
namespace Motor3D.Escena
{
	public class ElementoZBuffer : ObjetoEscena, IComparable<ElementoZBuffer>
	{

		private double[] mIndices;
		private int mNumeroIndices;

		private double mZ;
		public new event ModificadoEventHandler Modificado;
		public new delegate void ModificadoEventHandler(ref object Sender);

		public int NumeroIndices {
			get { return mNumeroIndices; }
		}

		public double[] Indices {
			get { return mIndices; }
			set {
				if (value.GetUpperBound(0) == mIndices.GetUpperBound(0)) {
					mIndices = value;
				}
			}
		}

		public double Indices {
			get {
				if (Index >= 0 && Index < mNumeroIndices) {
					return mIndices[Index];
				} else {
					throw new ExcepcionEscena("ELEMENTOZBUFFER (INDICE_GET): El índice está fuera del intervalo." + Constants.vbNewLine + "Index=" + Index.ToString() + Constants.vbNewLine + "Intervalo=(0," + mNumeroIndices - 1 + ")");
				}
			}
			set {
				if (Index >= 0 && Index < mNumeroIndices) {
					mIndices[Index] = value;
				} else {
					throw new ExcepcionEscena("ELEMENTOZBUFFER (INDICE_SET): El índice está fuera del intervalo." + Constants.vbNewLine + "Index=" + Index.ToString() + Constants.vbNewLine + "Intervalo=(0," + mNumeroIndices - 1 + ")");
				}
			}
		}

		public double Z {
			get { return mZ; }
			set {
				mZ = value;
				if (Modificado != null) {
					Modificado(this);
				}
			}
		}

		public object EsEquivalente(params double[] Indices)
		{
			return mIndices.Equals(Indices);
		}

		public ElementoZBuffer(int NumeroIndices)
		{
			if (NumeroIndices > 0) {
				mIndices = new double[NumeroIndices];
				mNumeroIndices = NumeroIndices;
			} else {
				throw new ExcepcionEscena("ELEMENTOZBUFFER (NEW): El número de índices debe ser mayor o igual que 1." + Constants.vbNewLine + "Numero de índices=" + NumeroIndices.ToString());
			}
		}

		public ElementoZBuffer(double Z, params double[] Indices)
		{
			mIndices = Indices;
			mNumeroIndices = Indices.GetUpperBound(0) + 1;
			mZ = Z;
		}

		public static bool operator >(ElementoZBuffer X, ElementoZBuffer Y)
		{
			return X.Z > Y.Z;
		}

		public static bool operator <(ElementoZBuffer X, ElementoZBuffer Y)
		{
			return X.Z < Y.Z;
		}

		public static bool operator ==(ElementoZBuffer X, ElementoZBuffer Y)
		{
			return X.Z == Y.Z;
		}

		public static bool operator !=(ElementoZBuffer X, ElementoZBuffer Y)
		{
			return X.Z != Y.Z;
		}

		public int CompareTo(ElementoZBuffer other)
		{
			switch (other.Z) {
				case  // ERROR: Case labels with binary operators are unsupported : GreaterThan
mZ:
					return 1;
				case  // ERROR: Case labels with binary operators are unsupported : Equality
mZ:
					return 0;
				case  // ERROR: Case labels with binary operators are unsupported : LessThan
mZ:
					return -1;
			}
		}
	}
}

File: motor3d 2.0/ElementoZBuffer.vb
Namespace Escena
    Public Class ElementoZBuffer
        Inherits ObjetoEscena
        Implements IComparable(Of ElementoZBuffer)

        Private mIndices() As Double
        Private mNumeroIndices As Integer
        Private mZ As Double

        Public Shadows Event Modificado(ByRef Sender As Object)

        Public ReadOnly Property NumeroIndices As Integer
            Get
                Return mNumeroIndices
            End Get
        End Property

        Public Property Indices() As Double()
            Get
                Return mIndices
            End Get
            Set(value As Double())
                If value.GetUpperBound(0) = mIndices.GetUpperBound(0) Then
                    mIndices = value
                End If
            End Set
        End Property

        Public Property Indices(ByVal Index As Integer) As Double
            Get
                If Index >= 0 AndAlso Index < mNumeroIndices Then
                    Return mIndices(Index)
                Else
                    Throw New ExcepcionEscena("ELEMENTOZBUFFER (INDICE_GET): El índice está fuera del intervalo." & vbNewLine _
                                              & "Index=" & Index.ToString & vbNewLine _
                                              & "Intervalo=(0," & mNumeroIndices - 1 & ")")
                End If
            End Get
            Set(value As Double)
                If Index >= 0 AndAlso Index < mNumeroIndices Then
                    mIndices(Index) = value
                Else
                    Throw New ExcepcionEscena("ELEMENTOZBUFFER (INDICE_SET): El índice está fuera del intervalo." & vbNewLine _
                          & "Index=" & Index.ToString & vbNewLine _
                          & "Intervalo=(0," & mNumeroIndices - 1 & ")")
                End If
            End Set
        End Property

        Public Property Z As Double
            Get
                Return mZ
            End Get
            Set(value As Double)
                mZ = value
                RaiseEvent Modificado(Me)
            End Set
        End Property

        Public Function EsEquivalente(ByVal ParamArray Indices() As Double)
            Return mIndices.Equals(Indices)
        End Function

        Public Sub New(ByVal NumeroIndices As Integer)
            If NumeroIndices > 0 Then
                ReDim mIndices(NumeroIndices - 1)
                mNumeroIndices = NumeroIndices
            Else
                Throw New ExcepcionEscena("ELEMENTOZBUFFER (NEW): El número de índices debe ser mayor o igual que 1." & vbNewLine _
                                          & "Numero de índices=" & NumeroIndices.ToString)
            End If
        End Sub

        Public Sub New(ByVal Z As Double, ByVal ParamArray Indices() As Double)
            mIndices = Indices
            mNumeroIndices = Indices.GetUpperBound(0) + 1
            mZ = Z
        End Sub

        Public Shared Operator >(ByVal X As ElementoZBuffer, ByVal Y As ElementoZBuffer) As Boolean
            Return X.Z > Y.Z
        End Operator

        Public Shared Operator <(ByVal X As ElementoZBuffer, ByVal Y As ElementoZBuffer) As Boolean
            Return X.Z < Y.Z
        End Operator

        Public Shared Operator =(ByVal X As ElementoZBuffer, ByVal Y As ElementoZBuffer) As Boolean
            Return X.Z = Y.Z
        End Operator

        Public Shared Operator <>(ByVal X As ElementoZBuffer, ByVal Y As ElementoZBuffer) As Boolean
            Return X.Z <> Y.Z
        End Operator

        Public Function CompareTo(other As ElementoZBuffer) As Integer Implements System.IComparable(Of ElementoZBuffer).CompareTo
            Select Case other.Z
                Case Is > mZ
                    Return 1
                Case Is = mZ
                    Return 0
                Case Is < mZ
                    Return -1
            End Select
        End Function
    End Class
End Namespace

